### PR TITLE
chore: Refactor VsTest related classes

### PR DESCRIPTION
**Pull request by dupdob**
This PR provides significant redesign for VsTest related classes, it also provides marginal improvements related to coverage analysis.
Planned features have been removed for now as they are more difficult to implement than anticipated.
But the redesign lead to clearer separations of concerns an will ease future evolution for this part of Stryker.
Plus, original ambition has not been abandoned yet and should be proposed through a future PR

- [x] refactoring of `VsTestRunner`, `VsTestRunnerPool` and `CoverageAnalyzer` classes to increase their consistency. Note that one additional class was introduced to improve the design '`VsTestContextInformation`
- ~[ ] ability to capture additional coverage info vs the original design via the usage of a JSON dump (WIP)~. _Exploratory testing failed to demonstrate this feature providing any benefit._ Related code will probably be removed for final version
- ~[x] smart coverage capture mode that relies on running coverage capture session for the problematic tests.~ This feature has been removed for now because it is not finished and this PR has been open for too long already.
- [x] a marginal gain in efficiency during testing phase via a finer selection of tests to run even in case of 'full test session'

### CODE REVIEW

**Comment by dupdob** [comment]
yoohoo: fixed all smells in one go.
BTW, any chance to fix the issue regarding the coverage gate (on sonar cloud) ? this is bothering because it appears as blocking whatever we do, plus it reports garbage metrics, If I understand this correctly

**Comment by rouke-broersma** [comment]
> yoohoo: fixed all smells in one go. BTW, any chance to fix the issue regarding the coverage gate (on sonar cloud) ? this is bothering because it appears as blocking whatever we do, plus it reports garbage metrics, If I understand this correctly

I agree, we should fix this.

File: .editorconfig [hunk 1]
```diff
@@ -1,5 +1,5 @@
-# top-most EditorConfig file
```
Comment (rouke-broersma) [inline .editorconfig:1]: Are the changes in this editorconfig supposed to be a part of this change?

**Comment by dupdob** [comment]
For info: AFAIC this PR is ready for review and merge. 
No further work planned, except for the occasional merge with main.

Awaiting for this one to be merged, as it will enable several other improvements in the pipeline

File: .editorconfig (hunk 1 above)
Comment (dupdob) [reply .editorconfig]: Some accident. Those changes have been reverted

**Comment by dupdob** [comment]
thanks for your time. First things first, I am looking into the unstable results....

File: docs/technical-reference/testing-framework.md [hunk 2]
```diff
@@ -27,39 +28,57 @@ This affects:
...
+During discovery phase, xUnit maps each data set to a test case; test case may be different (if they result in a new display name) or the same than an existing ne.
 if it can, those test cases are processed as usual; otherwise it will provide several results for the same test case.
 
-There is one test result per dataset, all associated with the same test case.
-
+It will then run all tests.
+### Run sequence
+Overall, xUnit first discovers all  test a
 Here is a summarized timeline of tests execution:
+xUnit runner calls data theory's data source to fetch each test case value(s)
```
[the remark this answers is not in the document]
Comment (dupdob) [reply docs/technical-reference/testing-framework.md:61]: yes

**Comment by dupdob** [comment]
It looks like VsTest crashes for some mutants. Still analyzing, but I suspect this may result in some random interpretation from Stryker.
This PR must not be merged until I fixes this.

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+        {
+            var options = new StrykerOptions();
+            var tests = new List<TestCase>
+            {
+                BuildCase("X0"),
+                BuildCase("X1")
+            };
+            var mockVsTest = BuildVsTestRunnerPool(options, out var runner, tests);
+            // assume 3 results for X0
+            SetupMockTestRun(mockVsTest, new[] { ("X0", true), ("X1", true), ("X0", true) , ("X0", true), ("X0", true)}, tests);
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs:360]: thanks: it was a temporary change while chasing a bug, and forgot to revert it

File: docs/technical-reference/testing-framework.md [hunk 4]
```diff
@@ -27,39 +28,57 @@ This affects:
...
 Run time theories may be discovered as one test, disregarding the number of underlying data set.
-During discovery phase, xUnit determines if it can generates a test case for each (theory) data set;
+During discovery phase, xUnit maps each data set to a test case; test case may be different (if they result in a new display name) or the same than an existing ne.
 if it can, those test cases are processed as usual; otherwise it will provide several results for the same test case.
 
-There is one test result per dataset, all associated with the same test case.
-
+It will then run all tests.
+### Run sequence
+Overall, xUnit first discovers all  test a
```
[the remark this answers is not in the document]
Comment (dupdob) [reply docs/technical-reference/testing-framework.md:59]: nope, but the text was erased by mistake, probably...

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+        public void ShouldComputeTimeoutProperly()
+        {
+            var mockVsTest = BuildVsTestRunnerPool(new StrykerOptions(), out var runner);
+            var now = DateTimeOffset.Now;
+            var duration = TimeSpan.FromMilliseconds(2);
+            var testResult = new TestResult(TestCases[0])
+            {
+                Outcome = TestOutcome.Passed,
+                StartTime = now,
+                EndTime = DateTimeOffset.Now+duration+duration,
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs:64]: I redesigned the test to make the intent more explicit: Stryker must use Duration as provided and should not try to recompute it via StartTime/EndTime

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs [hunk 6]
```diff
@@ -0,0 +1,498 @@
...
+                Outcome = TestOutcome.Passed,
+                StartTime = now,
+                EndTime = now+duration+duration,
+                Duration = duration
+            };
+            var otherTestResult = new TestResult(TestCases[0])
+            {
+                Outcome = TestOutcome.Passed,
+                StartTime = now,
+                EndTime = now+duration+duration,
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestRunnerPoolTests.cs:89]: ditto

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs [hunk 7]
```diff
@@ -0,0 +1,531 @@
...
+    private static void DiscoverTests(ITestDiscoveryEventsHandler discoveryEventsHandler, IReadOnlyCollection<TestCase> tests, bool aborted) =>
+        Task.Run(() => discoveryEventsHandler.HandleDiscoveredTests(tests)).
+            ContinueWith((_, u) => discoveryEventsHandler.HandleDiscoveryComplete((int)u, null, aborted), tests.Count);
+
+    protected TestCase BuildCase(string name, TestFramework framework = TestFramework.xUnit, string displayName = null)
+        => new(name, framework == TestFramework.xUnit ? _xUnitUri : _NUnitUri, _testAssemblyPath) { Id = new Guid(), DisplayName = displayName ?? name};
+
+    private TestCase FindOrBuildCase(string testResultId) => TestCases.FirstOrDefault(t => t.FullyQualifiedName == testResultId) ?? BuildCase(testResultId);
+
+    private static void MoqTestRun(ITestRunEventsHandler testRunEvents, IReadOnlyList<TestResult> testResults,
```
[the remark this answers is not in the document]
Comment (dupdob) [reply src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTestMockingHelper.cs:138]: done

**Comment by dupdob** [comment]
update: I fixed a minor issue and now I have quite stable results. See below for details.

Fixed problem: when there is only one mutation to test in a test session, it is activated as soon as possible (instead when the first covering test starts). This may result in wrong results such as survivor whereas it should be marked as Killed. This behavior was lost in my change, I restored it.

There is still a gap between head and this branch. This is due for two factors:
1) the order of tests has (probably) changed. This may trigger a switch between 'Timeout' and 'Killed' status: with NFluent, I have mutations that may trigger infinite recursion, hence stack overflow exceptions; as such, some tests fail and some others result in a stack overflow. With the current version, a killing test is run first, hence the mutant is marked as killed. With my PR, a 'stackoverflow' test is run first, resulting in a Timeout status (as of now, we can't distinguish between a crash and a normal timeout'
2) With this PR, concurrency is enabled when there is only one mutation to test in order to speed up things. Concurrency injects some randomness in the order of tests, which may result in variation between runs (Timeout and Killed).

(2) can be reverted, I am thinking about it.
(1) would be hard to revert, and I think we should not make such promise as ensuring the order of tests. Note that there is no guaranteed order by VsTest or the testrunners (in fact, xUnit reorders tests somewhat).

**Comment by rouke-broersma** [comment]
I agree that we should not guarantee the order in which unit tests are executed. Are mutation testruns within the same version of stryker also wildly inconsistent or is it only inconsistent between different version? I do think we should try to stay generally consistent within the same stryker version when running stryker multiple times against the same codebase (without guaranteeing).

**Comment by richardwerkman** [comment]
It seems I'm getting constistent results on this branch now. But it's still different from master :(

Mutants that were marked as survived are now marked as nocoverage. I checked and they should be covered!

Also I'm suddenly seeing these messages? 
image

I am running in multi test project mode on our integration test project. So that might be a starting point to look into this.

**Comment by dupdob** [comment]
There have been a slight change in how coverage is kept. But it should not result in that.
Any chance you can share this project?

File: src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTextContextInformationShould.cs [hunk 8]
```diff
@@ -24,15 +24,15 @@
 
 namespace Stryker.Core.UnitTest.TestRunners
 {
-    public class VsTextContextInformationShould
+    public class VsTextContextInformationTests
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core.UnitTest/TestRunners/VsTextContextInformationShould.cs:27]: Could you rename the file as well?

**Review by richardwerkman — APPROVED** [review]

**Comment by dupdob** [comment]
still some work to do:

- [x] fix the regression where the `Ignored` status was overwritten by `NoCoverage`
- [ ] add regression tests for this
- [ ] properly handle NUnit`s ignored tests (those are marked as `Skipped`)
- [ ] add associated tests

**Comment by dupdob** [comment]
👍

**Pull request merged**